games-board/megamek/megamek-0.34.10-r2.ebuild

```diff
@@ -1,21 +1,23 @@
-# Copyright 1999-2016 Gentoo Foundation
+# Copyright 1999-2021 Gentoo Authors
 # Distributed under the terms of the GNU General Public License v2
 
-EAPI=6
-inherit eutils java-pkg-2 java-ant-2
+EAPI=7
+inherit desktop edos2unix java-pkg-2 java-ant-2
 
-DESCRIPTION="an unofficial, online version of the Classic BattleTech board game"
+JAVA_PKG_WANT_SOURCE=1.5
+JAVA_PKG_WANT_TARGET=1.5
+
+DESCRIPTION="An unofficial, online version of the Classic BattleTech board game"
 HOMEPAGE="http://megamek.sourceforge.net/"
 SRC_URI="mirror://sourceforge/${PN}/MegaMek-v${PV}.zip"
 
 LICENSE="GPL-2"
 SLOT="0"
 KEYWORDS="~amd64 ~x86"
-IUSE=""
 
-DEPEND=">=virtual/jdk-1.5
-	app-arch/unzip"
-RDEPEND=">=virtual/jre-1.5"
+DEPEND="<virtual/jdk-1.9"
+RDEPEND="<virtual/jre-1.9"
```

Comment:
@fordfrog Here, I'm a bit uncertain of how to best designate the java dependencies: I tried compiling with openjdk-11 and it did not compile. With openjdk-8 it does compile. Thus, this is why I used the `(R)DEPEND` that you see here.

Is there a better way to do this?

Reply:
"it did not compile" is a bit vague. if it did not compile because you had source/target 1.5 then that's normal. you should start with `>=virtual/{jdk,jre}-1.8:*`, this will set source and target to 1.8. then check it with openjdk-8 and openjdk-11. if openjdk-11 fails, post the output. it might be easily fixable, or not :-)

Reply:
also, just a note, if something really does not compile with java 11 then the dep should be `virtual/jdk:1.8`. if something does not run with java 11, the dep should be `virtual/jre:1.8`.

Reply:
> also, just a note, if something really does not compile with java 11 then the dep should be `virtual/jdk:1.8`. if something does not run with java 11, the dep should be `virtual/jre:1.8`.

and also https://bugs.gentoo.org/787374 should be updated ...

Reply:
ok, so I tried this:

```diff
--- a/games-board/megamek/megamek-0.34.10-r2.ebuild
+++ b/games-board/megamek/megamek-0.34.10-r2.ebuild
@@ -4,8 +4,8 @@
 EAPI=7
 inherit desktop edos2unix java-pkg-2 java-ant-2
 
-JAVA_PKG_WANT_SOURCE=1.5
-JAVA_PKG_WANT_TARGET=1.5
+# JAVA_PKG_WANT_SOURCE=1.5
+# JAVA_PKG_WANT_TARGET=1.5
 
 DESCRIPTION="An unofficial, online version of the Classic BattleTech board game"
 HOMEPAGE="http://megamek.sourceforge.net/"
@@ -15,8 +15,8 @@ LICENSE="GPL-2"
 SLOT="0"
 KEYWORDS="~amd64 ~x86"
 
-DEPEND="<virtual/jdk-1.9"
-RDEPEND="<virtual/jre-1.9"
+DEPEND="virtual/jdk:1.8"
+RDEPEND="virtual/jre:1.8"
 BDEPEND="app-arch/unzip"
 
 S=${WORKDIR}
```

And this is the result I get:

<details>
<summary>build.log</summary>

```sh
[32;01m * [39;49;00mPackage:    games-board/megamek-0.34.10-r2
[32;01m * [39;49;00mRepository: gentoo
[32;01m * [39;49;00mMaintainer: games@gentoo.org
[32;01m * [39;49;00mUSE:        abi_x86_64 amd64 elibc_glibc kernel_linux userland_GNU
[32;01m * [39;49;00mFEATURES:   network-sandbox preserve-libs sandbox userpriv usersandbox
[32;01m * [39;49;00mPackage:    games-board/megamek-0.34.10-r2
[32;01m * [39;49;00mRepository: gentoo
[32;01m * [39;49;00mMaintainer: games@gentoo.org
[32;01m * [39;49;00mUSE:        abi_x86_64 amd64 elibc_glibc kernel_linux userland_GNU
[32;01m * [39;49;00mFEATURES:   network-sandbox preserve-libs sandbox userpriv usersandbox
 [32;01m*[0m Using: openjdk-bin-8
Unable to configure loopback interface: Operation not permitted
>>> Unpacking source...
>>> Unpacking MegaMek-v0.34.10.zip to /var/tmp/portage/games-board/megamek-0.34.10-r2/work
>>> Source unpacked in /var/tmp/portage/games-board/megamek-0.34.10-r2/work
Unable to configure loopback interface: Operation not permitted
>>> Preparing source in /var/tmp/portage/games-board/megamek-0.34.10-r2/work ...
removed 'MegaMek.jar'
>>> Source prepared.
Unable to configure loopback interface: Operation not permitted
>>> Configuring source in /var/tmp/portage/games-board/megamek-0.34.10-r2/work ...
Rewriting attributes
Rewriting ./build.xml
>>> Source configured.
Unable to configure loopback interface: Operation not permitted
>>> Compiling source in /var/tmp/portage/games-board/megamek-0.34.10-r2/work ...
 [32;01m*[0m Disabling all optional ANT_TASKS
Buildfile: /var/tmp/portage/games-board/megamek-0.34.10-r2/work/build.xml

clean:
   [delete] Deleting: /var/tmp/portage/games-board/megamek-0.34.10-r2/work/logs/timestamp

init:
    [mkdir] Created dir: /var/tmp/portage/games-board/megamek-0.34.10-r2/work/classes

compile:
    [javac] Compiling 1334 source files to /var/tmp/portage/games-board/megamek-0.34.10-r2/work/classes
    [javac] /var/tmp/portage/games-board/megamek-0.34.10-r2/work/src/megamek/client/ui/AWT/boardview3d/ArrowModel.java:5: error: unmappable character for encoding UTF8
    [javac]  * This file (C) 2008 J�rg Walter <j.walter@syntax-k.de>
    [javac]                        ^
    [javac] /var/tmp/portage/games-board/megamek-0.34.10-r2/work/src/megamek/client/ui/AWT/boardview3d/ArtilleryAttackModel.java:5: error: unmappable character for encoding UTF8
    [javac]  * This file (C) 2008 J�rg Walter <j.walter@syntax-k.de>
    [javac]                        ^
    [javac] /var/tmp/portage/games-board/megamek-0.34.10-r2/work/src/megamek/client/ui/AWT/boardview3d/AttackGroup.java:5: error: unmappable character for encoding UTF8
    [javac]  * This file (C) 2008 J�rg Walter <j.walter@syntax-k.de>
    [javac]                        ^
    [javac] /var/tmp/portage/games-board/megamek-0.34.10-r2/work/src/megamek/client/ui/AWT/boardview3d/AttackModel.java:5: error: unmappable character for encoding UTF8
    [javac]  * This file (C) 2008 J�rg Walter <j.walter@syntax-k.de>
    [javac]                        ^
    [javac] /var/tmp/portage/games-board/megamek-0.34.10-r2/work/src/megamek/client/ui/AWT/boardview3d/BoardHexModel.java:5: error: unmappable character for encoding UTF8
    [javac]  * This file (C) 2008 J�rg Walter <j.walter@syntax-k.de>
    [javac]                        ^
    [javac] /var/tmp/portage/games-board/megamek-0.34.10-r2/work/src/megamek/client/ui/AWT/boardview3d/BoardModel.java:5: error: unmappable character for encoding UTF8
    [javac]  * This file (C) 2008 J�rg Walter <j.walter@syntax-k.de>
    [javac]                        ^
    [javac] /var/tmp/portage/games-board/megamek-0.34.10-r2/work/src/megamek/client/ui/AWT/boardview3d/BoardView3D.java:5: error: unmappable character for encoding UTF8
    [javac]  * This file also (C) 2008 J�rg Walter <j.walter@syntax-k.de>
    [javac]                             ^
    [javac] /var/tmp/portage/games-board/megamek-0.34.10-r2/work/src/megamek/client/ui/AWT/boardview3d/C.java:5: error: unmappable character for encoding UTF8
    [javac]  * This file (C) 2008 J�rg Walter <j.walter@syntax-k.de>
    [javac]                        ^
    [javac] /var/tmp/portage/games-board/megamek-0.34.10-r2/work/src/megamek/client/ui/AWT/boardview3d/C3LinkModel.java:5: error: unmappable character for encoding UTF8
    [javac]  * This file (C) 2008 J�rg Walter <j.walter@syntax-k.de>
    [javac]                        ^
    [javac] /var/tmp/portage/games-board/megamek-0.34.10-r2/work/src/megamek/client/ui/AWT/boardview3d/ConnectionModel.java:5: error: unmappable character for encoding UTF8
    [javac]  * This file (C) 2008 J�rg Walter <j.walter@syntax-k.de>
    [javac]                        ^
    [javac] /var/tmp/portage/games-board/megamek-0.34.10-r2/work/src/megamek/client/ui/AWT/boardview3d/CursorModel.java:5: error: unmappable character for encoding UTF8
    [javac]  * This file (C) 2008 J�rg Walter <j.walter@syntax-k.de>
    [javac]                        ^
    [javac] /var/tmp/portage/games-board/megamek-0.34.10-r2/work/src/megamek/client/ui/AWT/boardview3d/EntityGroup.java:5: error: unmappable character for encoding UTF8
    [javac]  * This file (C) 2008 J�rg Walter <j.walter@syntax-k.de>
    [javac]                        ^
    [javac] /var/tmp/portage/games-board/megamek-0.34.10-r2/work/src/megamek/client/ui/AWT/boardview3d/EntityModel.java:5: error: unmappable character for encoding UTF8
    [javac]  * This file (C) 2008 J�rg Walter <j.walter@syntax-k.de>
    [javac]                        ^
    [javac] /var/tmp/portage/games-board/megamek-0.34.10-r2/work/src/megamek/client/ui/AWT/boardview3d/HexModel.java:5: error: unmappable character for encoding UTF8
    [javac]  * This file (C) 2008 J�rg Walter <j.walter@syntax-k.de>
    [javac]                        ^
    [javac] /var/tmp/portage/games-board/megamek-0.34.10-r2/work/src/megamek/client/ui/AWT/boardview3d/ImageModel.java:5: error: unmappable character for encoding UTF8
    [javac]  * This file (C) 2008 J�rg Walter <j.walter@syntax-k.de>
    [javac]                        ^
    [javac] /var/tmp/portage/games-board/megamek-0.34.10-r2/work/src/megamek/client/ui/AWT/boardview3d/LabelModel.java:5: error: unmappable character for encoding UTF8
    [javac]  * This file (C) 2008 J�rg Walter <j.walter@syntax-k.de>
    [javac]                        ^
    [javac] /var/tmp/portage/games-board/megamek-0.34.10-r2/work/src/megamek/client/ui/AWT/boardview3d/MoveGroup.java:5: error: unmappable character for encoding UTF8
    [javac]  * This file (C) 2008 J�rg Walter <j.walter@syntax-k.de>
    [javac]                        ^
    [javac] /var/tmp/portage/games-board/megamek-0.34.10-r2/work/src/megamek/client/ui/AWT/boardview3d/MoveStepModel.java:5: error: unmappable character for encoding UTF8
    [javac]  * This file (C) 2008 J�rg Walter <j.walter@syntax-k.de>
    [javac]                        ^
    [javac] /var/tmp/portage/games-board/megamek-0.34.10-r2/work/src/megamek/client/ui/AWT/boardview3d/SignModel.java:5: error: unmappable character for encoding UTF8
    [javac]  * This file (C) 2008 J�rg Walter <j.walter@syntax-k.de>
    [javac]                        ^
    [javac] /var/tmp/portage/games-board/megamek-0.34.10-r2/work/src/megamek/client/ui/AWT/boardview3d/TileTextureManager.java:5: error: unmappable character for encoding UTF8
    [javac]  * This file (C) 2008 J�rg Walter <j.walter@syntax-k.de>
    [javac]                        ^
    [javac] /var/tmp/portage/games-board/megamek-0.34.10-r2/work/src/megamek/client/ui/AWT/boardview3d/ViewTransform.java:5: error: unmappable character for encoding UTF8
    [javac]  * This file (C) 2008 J�rg Walter <j.walter@syntax-k.de>
    [javac]                        ^
    [javac] /var/tmp/portage/games-board/megamek-0.34.10-r2/work/src/megamek/client/ui/IDisplayable.java:5: error: unmappable character for encoding UTF8
    [javac]  * This file (C) 2008 J�rg Walter <j.walter@syntax-k.de>
    [javac]                        ^
    [javac] /var/tmp/portage/games-board/megamek-0.34.10-r2/work/src/megamek/client/ui/ITilesetManager.java:5: error: unmappable character for encoding UTF8
    [javac]  * This file (C) 2008 J�rg Walter <j.walter@syntax-k.de>
    [javac]                        ^
    [javac] 23 errors

BUILD FAILED
/var/tmp/portage/games-board/megamek-0.34.10-r2/work/build.xml:59: Compile failed; see the compiler error output for details.

Total time: 2 seconds
 [31;01m*[0m ERROR: games-board/megamek-0.34.10-r2::gentoo failed (compile phase):
 [31;01m*[0m   eant failed
 [31;01m*[0m 
 [31;01m*[0m Call stack:
 [31;01m*[0m     ebuild.sh, line  125:  Called src_compile
 [31;01m*[0m   environment, line 3289:  Called eant
 [31;01m*[0m   environment, line  571:  Called die
 [31;01m*[0m The specific snippet of code:
 [31;01m*[0m       ant ${antflags} "${@}" || die "eant failed"
 [31;01m*[0m 
 [31;01m*[0m If you need support, post the output of `emerge --info '=games-board/megamek-0.34.10-r2::gentoo'`,
 [31;01m*[0m the complete build log and the output of `emerge -pqv '=games-board/megamek-0.34.10-r2::gentoo'`.
!!! When you file a bug report, please include the following information:
GENTOO_VM=openjdk-bin-8  CLASSPATH="" JAVA_HOME="/opt/openjdk-bin-8.292_p10"
JAVACFLAGS="-source 1.8 -target 1.8" COMPILER="javac"
and of course, the output of emerge --info =megamek-0.34.10
 [31;01m*[0m The complete build log is located at '/var/tmp/portage/games-board/megamek-0.34.10-r2/temp/build.log'.
 [31;01m*[0m The ebuild environment file is located at '/var/tmp/portage/games-board/megamek-0.34.10-r2/temp/environment'.
 [31;01m*[0m Working directory: '/var/tmp/portage/games-board/megamek-0.34.10-r2/work'
 [31;01m*[0m S: '/var/tmp/portage/games-board/megamek-0.34.10-r2/work'
```

</details>

If I change the `(R)DEPEND` back to `>=virtual/j{dk,re}-1.5`, I can get it to compile with `openjdk-bin-8`. If I try with `openjdk-bin-11`, I get the following:

<details>
<summary> build.log</summary>

[32;01m * [39;49;00mPackage:    games-board/megamek-0.34.10-r2
[32;01m * [39;49;00mRepository: gentoo
[32;01m * [39;49;00mMaintainer: games@gentoo.org
[32;01m * [39;49;00mUSE:        abi_x86_64 amd64 elibc_glibc kernel_linux userland_GNU
[32;01m * [39;49;00mFEATURES:   network-sandbox preserve-libs sandbox userpriv usersandbox
[32;01m * [39;49;00mPackage:    games-board/megamek-0.34.10-r2
[32;01m * [39;49;00mRepository: gentoo
[32;01m * [39;49;00mMaintainer: games@gentoo.org
[32;01m * [39;49;00mUSE:        abi_x86_64 amd64 elibc_glibc kernel_linux userland_GNU
[32;01m * [39;49;00mFEATURES:   network-sandbox preserve-libs sandbox userpriv usersandbox
 [32;01m*[0m Using: openjdk-bin-11
Unable to configure loopback interface: Operation not permitted
>>> Unpacking source...
>>> Unpacking MegaMek-v0.34.10.zip to /var/tmp/portage/games-board/megamek-0.34.10-r2/work
>>> Source unpacked in /var/tmp/portage/games-board/megamek-0.34.10-r2/work
Unable to configure loopback interface: Operation not permitted
>>> Preparing source in /var/tmp/portage/games-board/megamek-0.34.10-r2/work ...
removed 'MegaMek.jar'
>>> Source prepared.
Unable to configure loopback interface: Operation not permitted
>>> Configuring source in /var/tmp/portage/games-board/megamek-0.34.10-r2/work ...
Rewriting attributes
Rewriting ./build.xml
>>> Source configured.
Unable to configure loopback interface: Operation not permitted
>>> Compiling source in /var/tmp/portage/games-board/megamek-0.34.10-r2/work ...
 [32;01m*[0m Disabling all optional ANT_TASKS
Warning: Unable to determine tools.jar location.
  If build fails because sun.* classes could not be found,
  Make sure you are using a JDK, not JRE as your user/system VM.
  and that you have java-config version 2.0.30 or above installed.
Buildfile: /var/tmp/portage/games-board/megamek-0.34.10-r2/work/build.xml

clean:
   [delete] Deleting: /var/tmp/portage/games-board/megamek-0.34.10-r2/work/logs/timestamp

init:
    [mkdir] Created dir: /var/tmp/portage/games-board/megamek-0.34.10-r2/work/classes

compile:
    [javac] Compiling 1334 source files to /var/tmp/portage/games-board/megamek-0.34.10-r2/work/classes
    [javac] error: Source option 5 is no longer supported. Use 6 or later.
    [javac] error: Target option 1.5 is no longer supported. Use 1.6 or later.

BUILD FAILED
/var/tmp/portage/games-board/megamek-0.34.10-r2/work/build.xml:59: Compile failed; see the compiler error output for details.

Total time: 1 second
 [31;01m*[0m ERROR: games-board/megamek-0.34.10-r2::gentoo failed (compile phase):
 [31;01m*[0m   eant failed
 [31;01m*[0m 
 [31;01m*[0m Call stack:
 [31;01m*[0m     ebuild.sh, line  125:  Called src_compile
 [31;01m*[0m   environment, line 3289:  Called eant
 [31;01m*[0m   environment, line  571:  Called die
 [31;01m*[0m The specific snippet of code:
 [31;01m*[0m       ant ${antflags} "${@}" || die "eant failed"
 [31;01m*[0m 
 [31;01m*[0m If you need support, post the output of `emerge --info '=games-board/megamek-0.34.10-r2::gentoo'`,
 [31;01m*[0m the complete build log and the output of `emerge -pqv '=games-board/megamek-0.34.10-r2::gentoo'`.
!!! When you file a bug report, please include the following information:
GENTOO_VM=openjdk-bin-11  CLASSPATH="" JAVA_HOME="/opt/openjdk-bin-11.0.11_p9"
JAVACFLAGS="-source 1.5 -target 1.5" COMPILER="javac"
and of course, the output of emerge --info =megamek-0.34.10
 [31;01m*[0m The complete build log is located at '/var/tmp/portage/games-board/megamek-0.34.10-r2/temp/build.log'.
 [31;01m*[0m The ebuild environment file is located at '/var/tmp/portage/games-board/megamek-0.34.10-r2/temp/environment'.
 [31;01m*[0m Working directory: '/var/tmp/portage/games-board/megamek-0.34.10-r2/work'
 [31;01m*[0m S: '/var/tmp/portage/games-board/megamek-0.34.10-r2/work'

</details>

So based on these tests, I concluded:

1. It **will** compile with jdk-8
2. It **will not** compile with jdk-11
3. It needs to **think** it's using jdk-1.5 (or w/e, I dunno)

Number three is what led me to adding `JAVA_PKG_WANT_{SOURCE,TARGET}`, which allows me to update the `(R)DEPEND` to a valid version (the `1.5` gives a QA warning) while still allowing the package to compile.

Reply:
> So based on these tests, I concluded:
> 
>     1. It **will** compile with jdk-8
> 
>     2. It **will not** compile with jdk-11
> 
>     3. It needs to **think** it's using jdk-1.5 (or w/e, I dunno)
> 
> 
> Number three is what led me to adding `JAVA_PKG_WANT_{SOURCE,TARGET}`, which allows me to update the `(R)DEPEND` to a valid version (the `1.5` gives a QA warning) while still allowing the package to compile.

The [bug](https://bugs.gentoo.org/787374) assignee might decide to either last-rite, upgrade or fix it.

Reply:
@ezzieyguywuf it's obviously problem with char code page. you need to set `JAVA_ANT_ENCODING="ISO-8859-1"`, that should fix the issue with openjdk-8 and if you keep min version at 1.8, it will also compile with openjdk-11 unless there are other issues.

Reply:
> it's obviously problem with char code page

Yes, I agree this is obvious :-)

> you need to set `JAVA_ANT_ENCODING="ISO-8859-1"`

This solution is not as obvious, at least for someone less familiar with this java stuff (like me! 🤣)

> That should fix the issue with openjdk-8

I will test it out and report back here

> it will also compile with openjdk-11

I believe I already tested this, and it does **not** in fact compile with openjdk-11, but I will test again with the character encoding fix and see what happens.

Do I need to perform runtime testing too to ensure the openjdk-11 stuff works as expected?

Reply:
Hah, it works!

> Do I need to perform runtime testing too to ensure the openjdk-11 stuff works as expected?

this works too